src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    public sealed class FilterForUpload : Task
+    {
+        /// <summary>
+        /// This task is used to select the appropriate test zips from the different output directories produced by the build,
+        /// which are created based on OSGroup and TestTFM.
+        /// </summary>
+        [Required]
+        public ITaskItem[] UnfilteredUploadItems { get; set; }
+
+        /// <summary>
+        /// This is the TestTFM folder that we want to select test zips from. This needs to be on the path to the test zip.
+        /// </summary>
+        [Required]
+        public string FilterToTestTFM { get; set; }
+
+        /// <summary>
+        /// This is the OSGroup folder that we want to select test zips from. This does not require an exact match as some tests 
+        /// may or may not have an exact OSGroup folder match.
+        /// </summary>
+        [Required]
+        public string FilterToOSGroup { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        private readonly List<ITaskItem> FilteredUploadItemsList = new List<ITaskItem>();
+
+        public override bool Execute()
+        {
+           
+            List < ITaskItem > temp = new List<ITaskItem>();
+            
+            foreach (var item in UnfilteredUploadItems)
+            {
+                if (item.GetMetadata("TestTFM").Contains(FilterToTestTFM))
+                {
+                    temp.Add(item);
+                }
+            }
+
+            Dictionary<string, List<ITaskItem>> map = new Dictionary<string, List<ITaskItem>>();
+            foreach (var item in temp)
+            {
+                string testName = Path.GetFileName(item.ItemSpec);
+                List<ITaskItem> values = null;
+                if (map.ContainsKey(testName))
+                {
+                    values = null;
+                    map.TryGetValue(testName, out values);
+                }
+                else
+                {
+                    values = new List<ITaskItem>();
+                }
+                values.Add(item);
+                map[testName] = values;
```

Comment:
You shouldn't have to set this value except in the else (since it's already in the dictionary)